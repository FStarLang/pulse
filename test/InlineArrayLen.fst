module InlineArrayLen

#lang-pulse
open Pulse

fn basic ()
  returns int
{
  let mut arr = [| 123; 2sz |];
  arr.(0sz);
}

inline_for_extraction noextract
fn gen (x : SizeT.t)
  requires pure (SizeT.v x > 0)
  returns int
{
  let mut arr = [| 123; x |];
  arr.(0sz);
}

fn use ()
  returns int
{
  gen 2sz;
}

inline_for_extraction noextract
fn gen_init (x : SizeT.t)
  (init : unit -> int)
  requires pure (SizeT.v x > 0)
  returns int
{
  let mut arr = [| init(); 2sz |];
  let r = arr.(0sz);
  r
}

fn use_gen_init ()
  returns int
{
  gen_init 2sz (fun _ -> 123);
}

inline_for_extraction noextract
fn gen_init_st (x : SizeT.t)
  (init : unit -> stt int emp (fun _ -> emp))
  requires pure (SizeT.v x > 0)
  returns int
{
  let mut arr = [| init(); 2sz |];
  let r = arr.(0sz);
  r
}

fn use_gen_init_st ()
  returns int
{
  fn init ()
    returns int
  {
    42;
  };
  gen_init_st 2sz init;
}

inline_for_extraction noextract
fn gen_len
  (len : (unit -> x:SizeT.t{SizeT.v x > 0}))
  returns int
{
  let mut arr = [| 123; len() |];
  let r = arr.(0sz);
  r
}

fn use_gen_len ()
  returns int
{
  gen_len (fun _ -> 2sz);
}

inline_for_extraction noextract
fn gen_len_st
  (len : unit -> stt SizeT.t emp (fun r -> pure (SizeT.v r > 0)))
  returns int
{
  let mut arr = [| 123; len () |];
  let r = arr.(0sz);
  r
}

fn use_gen_len_st ()
  returns int
{
  fn len ()
    returns r : SizeT.t
    ensures pure (SizeT.v r > 0)
  {
    42sz;
  };
  gen_len_st len;
}