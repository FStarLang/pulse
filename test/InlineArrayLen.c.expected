/* krml header omitted for test repeatability */


#include "InlineArrayLen.h"

krml_checked_int_t InlineArrayLen_basic(void)
{
  krml_checked_int_t arr[2U];
  for (uint32_t _i = 0U; _i < (size_t)2U; ++_i)
    arr[_i] = (krml_checked_int_t)123;
  return arr[0U];
}

krml_checked_int_t InlineArrayLen_use(void)
{
  krml_checked_int_t arr[2U];
  for (uint32_t _i = 0U; _i < (size_t)2U; ++_i)
    arr[_i] = (krml_checked_int_t)123;
  return arr[0U];
}

krml_checked_int_t InlineArrayLen_use_gen_init(void)
{
  krml_checked_int_t arr[2U];
  for (uint32_t _i = 0U; _i < (size_t)2U; ++_i)
    arr[_i] = (krml_checked_int_t)123;
  return arr[0U];
}

krml_checked_int_t InlineArrayLen_use_gen_init_st(void)
{
  krml_checked_int_t __anf = (krml_checked_int_t)42;
  krml_checked_int_t arr[2U];
  for (uint32_t _i = 0U; _i < (size_t)2U; ++_i)
    arr[_i] = __anf;
  return arr[0U];
}

krml_checked_int_t InlineArrayLen_use_gen_len(void)
{
  krml_checked_int_t arr[2U];
  for (uint32_t _i = 0U; _i < (size_t)2U; ++_i)
    arr[_i] = (krml_checked_int_t)123;
  return arr[0U];
}

krml_checked_int_t InlineArrayLen_use_gen_len_st(void)
{
  size_t __anf = (size_t)42U;
  KRML_CHECK_SIZE(sizeof (krml_checked_int_t), __anf);
  krml_checked_int_t arr[__anf];
  for (uint32_t _i = 0U; _i < __anf; ++_i)
    arr[_i] = (krml_checked_int_t)123;
  return arr[0U];
}

